# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

## 读取数据

In [13]:
# 加载包
import pandas as pd 
import numpy as np

# 加载数据
original_titles = pd.read_csv('E:/file/data_analysis/python/小白玩转python/随课数据/数据整理篇/titles.csv')
original_credits = pd.read_csv('E:/file/data_analysis/python/小白玩转python/随课数据/数据整理篇/credits.csv')

In [2]:
original_titles.head(5)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

## 评估数据和清理数据

在这里将会对`original_titles`和`original_credits`  DataFrame所包含的数据进行评估和清理。  
主要从两个方面进行：结构和内容，即整齐度与干净度。  
数据的结构性问题指的是不符合“每个变量为一列，每个观察值为一行，每个类型的观察单位为一个表格”这三个标准；数据的内容型问题包括存在丢失数据、重复数据、无效数据等。

为了区分清理前后的数据，把清理前的数据命名为`original_titles`和`original_credits` ，把清理后的数据命名为`cleaned_titles`和`cleaned_credits`。

### 数据整齐度

In [66]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

In [67]:
cleaned_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [68]:
# 展示数据更多细节
# pd.set_option('display.max_columns',150) # 扩展展示列的上限
# pd.set_option('display.max_colwidth',500)# 扩展每个列的字符显示上限
cleaned_titles.sample(10)

id                                  title   type  \
1883    ts86296                  Sunderland 'Til I Die   SHOW   
2837   tm445258                        Marriage Palace  MOVIE   
3780   tm816992                           Double World  MOVIE   
4987   ts310938                            New Heights   SHOW   
3402   ts192440          Garth Brooks: The Road I'm On   SHOW   
1929   tm233294                            War Machine  MOVIE   
1579   tm243931                  The Beginning of Life  MOVIE   
678    tm167232                                Unknown  MOVIE   
2952  tm1073433  Shiva: The Secret World Of Vedas City  MOVIE   
196    tm128073                 Bill Hicks: Relentless  MOVIE   

                                            description  release_year  \
1883  This docuseries follows English soccer club Su...          2018   
2837  Marriage Palace is a comedy movie which takes ...          2018   
3780  Seeing the neighboring country become more and...          2020   
4987  When business consultant Michi Wyss inherits h...          2021   
3402  A look at the life and career of Garth Brooks ...          2019   
1929  A rock star general bent on winning the “impos...          2017   
1579  One of the greatest neuroscience breakthroughs...          2016   
678   A man awakens from a coma, only to discover th...          2011   
2952  When the residents of Vedas are ordered to eva...          2017   
196   Bill Hicks tells us how he feels about non-smo...          1992   

     age_certification  runtime                              genres  \
1883             TV-MA       41          ['documentation', 'sport']   
2837               NaN      102                          ['comedy']   
3780             PG-13      110               ['fantasy', 'action']   
4987             TV-MA       48                           ['drama']   
3402             TV-PG       86        ['documentation', 'reality']   
1929                 R      122          ['drama', 'comedy', 'war']   
1579                 G       98         ['documentation', 'family']   
678              PG-13      113  ['thriller', 'action', 'european']   
2952                 G       78                                  []   
196                NaN       61                          ['comedy']   

     production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
1883               ['GB']      2.0   tt8914684         8.1      8113.0   
2837               ['XX']      NaN   tt8852242         6.2       184.0   
3780         ['CN', 'HK']      NaN  tt10508838         6.1      4160.0   
4987               ['CH']      1.0  tt15393644         6.7       525.0   
3402               ['US']      1.0  tt10095402         7.4       287.0   
1929               ['US']      NaN   tt4758646         6.0     46276.0   
1579               ['BR']      NaN   tt5140248         7.5       940.0   
678    ['GB', 'FR', 'DE']      NaN   tt1401152         6.8    260783.0   
2952                   []      NaN         NaN         NaN         NaN   
196                ['GB']      NaN         NaN         NaN         NaN   

      tmdb_popularity  tmdb_score  
1883            5.099       8.000  
2837            0.840       7.000  
3780           14.660       6.700  
4987            8.664       7.900  
3402            0.699      10.000  
1929           12.235       5.552  
1579            1.134       7.700  
678            63.884       6.700  
2952            1.400         NaN  
196             3.822       7.800

从随机抽取的10个数据来看,`cleaned_titles`的`genres	`和`production_countries`的变量中包含多个值，应当进行拆分  
先取一个`genres	`变量进行观察

In [69]:
cleaned_titles['genres'][1]

"['drama', 'crime']"

虽然`genres`表示形式是列表，但其实并非字符串列表,而是字符串，无法直接使用`value_counts()`统计各个值出现的次数，我们可以使用python内置的`eval`函数，它可以把字符串转化成表达式，所以可以帮助我们把表示列表的字符串转成列表本身。

In [70]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(lambda s: eval(s))
cleaned_titles['genres'][1]

['drama', 'crime']

转化为表达式后，就可使用`DataFrame`的`explode`方法，把列表拆成单独的行

In [71]:
cleaned_titles = cleaned_titles.explode("genres")
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

后续对`production_countries`也是同样的流程，每个观察值的`production_countries`值并不是单个国家，先提取任意一个`production_countries`变量的一个值进行观察

In [72]:
cleaned_titles['production_countries'][20]

20    ['EG']
20    ['EG']
20    ['EG']
20    ['EG']
20    ['EG']
Name: production_countries, dtype: object

可以看到`production_countries`也是同样的问题，虽然表示形式为列表，但其实际类型并非字符串列表，而是字符串，难以进行拆分。我们再次利用`eval`函数进行转换，并确认转换后的数据类型

In [73]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda s: eval(s))
cleaned_titles['production_countries'][1]

1    [US]
1    [US]
Name: production_countries, dtype: object

确认类型转换完毕，利用`explode`方法，把列表值拆分成单独的行

In [74]:
cleaned_titles = cleaned_titles.explode("production_countries")
cleaned_titles.sample(10)

id                                 title   type  \
1767   tm414073                             Backtrace  MOVIE   
3718   ts251615                              (Un)Well   SHOW   
392    tm134235               One Piece: Strong World  MOVIE   
1188    ts36177                                Glitch   SHOW   
4364   ts225708             Six Windows in the Desert   SHOW   
2340    ts76511       The Big Family Cooking Showdown   SHOW   
5410  tm1000599                       The Last Forest  MOVIE   
3585    ts86365                   Rilakkuma and Kaoru   SHOW   
3139  tm1121246                                Code 8  MOVIE   
5211   ts320426  Crime Scene: The Times Square Killer   SHOW   

                                            description  release_year  \
1767  The lone surviving thief of a violent armored ...          2018   
3718  This docuseries takes a deep dive into the luc...          2020   
392   20 years after his escape from Impel Down, the...          2009   
1188  A police officer and a doctor face an emotiona...          2015   
4364  From a theatre play to the aftermath of a plan...          2020   
2340  In this unscripted series, families passionate...          2017   
5410  In powerful images, alternating between docume...          2021   
3585  Kaoru's unexpected new roommate is Rilakkuma, ...          2019   
3139  In Lincoln City, some inhabitants have extraor...          2019   
5211  In 1970s NYC, the “Torso Killer” preys on wome...          2021   

     age_certification  runtime         genres production_countries  seasons  \
1767                 R       92       thriller                   CA      NaN   
3718             TV-MA       52  documentation                   US      1.0   
392              PG-13      114      animation                   JP      NaN   
1188             TV-MA       52         horror                   AU      3.0   
4364             TV-14       21          scifi                   SA      1.0   
2340             TV-PG       59       european                   GB      2.0   
5410               NaN       74          drama                   BR      NaN   
3585             TV-PG       12         family                   JP      1.0   
3139               NaN       99          drama                   CA      NaN   
5211             TV-MA       49       thriller                   US      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1767   tt3588588         3.8      4453.0           19.984       5.400  
3718  tt12759384         6.4      1068.0            3.075       6.700  
392    tt1485763         7.5      5467.0           39.946       7.500  
1188   tt4192782         7.3     17240.0           48.448       6.800  
4364  tt11874042         4.3      3104.0            1.176      10.000  
2340   tt7518558         7.0       682.0            2.706       7.400  
5410  tt14029622         7.3       425.0            2.176       7.000  
3585   tt9348716         8.2      1616.0            8.237       7.400  
3139   tt6259380         6.1     39469.0           18.193       6.200  
5211  tt16027926         6.5      3974.0            5.848       7.125

在处理完`cleaned_titiles`结构性问题后，查看`cleaned_credits`

In [75]:
cleaned_credits.sample(10)

person_id         id              name            character   role
36780     746513   tm365122  Kirk H. Andersen           Pedestrian  ACTOR
71058     194552   tm473502     Satish Alekar    Sheshrao Wankhede  ACTOR
36300    1828135    ts79086        Tony Garza  Jose Trinidad Marin  ACTOR
41450     807574   tm445153       Anju Kurian              Shruthi  ACTOR
23970      56167   tm219173   Brian F. Durkin          John Stasio  ACTOR
35121    1156529    ts88234         Kian Hunt    Azhaar's Gangster  ACTOR
36519      17287   tm425884      Alan Cumming     Narrator (voice)  ACTOR
5952      984522    tm70807       Anika Ellis             Infected  ACTOR
75257     129208  tm1195428   Ernesto Sevilla                  NaN  ACTOR
40280     254823   tm244277       Stunt Silva          Bullet Raja  ACTOR

从随机的10个数据来看，`cleaned_credits`数据符合“每个变量为一列，每个观察值为一行，每个类型的的观察单位为一个表格”因此不存在结构性问题

### 数据干净度

接下来使用`info()`函数，对数据内容进行大致了解

In [76]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

从输出结果来看`cleaned_titles`数据共有17818条观察值，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量均存在缺失值，将在后续进行评估和清理。  
此外此外，`release_year`表示年份，数据类型不应为数字，应为日期，所以需要进行数据格式转换。

In [77]:
cleaned_titles['release_year'] = pd.to_datetime(cleaned_titles['release_year'],format='%Y')
cleaned_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   

                                         description release_year  \
0  This collection includes 12 World War II-era p...   1945-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300

In [78]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


从输出结果来看，`cleaned_credits`数据共有77801行，其中`character`变量有缺失，将在后续进行评估和清理。  
此外，`person_id`为演员ID，数据类型应为字符串，所以需要进行格式转换。

In [79]:
cleaned_credits['person_id'] = cleaned_credits['person_id'].astype(str)
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  object
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: object(5)
memory usage: 3.0+ MB


#### 处理缺失数据

在`cleaned_titles`中，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量存在缺失值。  

由于影视作品的标题、描述、适龄认证、发行国家、电视剧季数，IMDB的ID、TMDB的流行度、TMDB的评分，并不影响我们挖掘各个流派中高IMDB评分的作品的演员，所以可以保留`title`、`description`、`age_certification`、`production_countries`、`seasons`、`imdb_id`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量中存在的空缺观察值。  
但是`imdb_score`和`genres`，即IMDB评分和流派，和我们的后续分析息息相关。  
先提取出`imdb_score`缺失观察值进行查看。

In [80]:
cleaned_titles.query('imdb_score.isnull()')

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

由于缺失分析所需的核心数据，因此把这些观察值进行删除，并检查删除后该列空缺值的个数

In [81]:
cleaned_titles = cleaned_titles.dropna(subset='imdb_score')
cleaned_titles['imdb_score'].isnull().sum()

0

然后提取`genres`缺失观察值进行查看

In [82]:
cleaned_titles.query('genres.isnull()')

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

由于缺失分析所需的核心数据`genres`,所以把这些观察值删除，并检查删除后的空缺值个数

In [83]:
cleaned_titles = cleaned_titles.dropna(subset='genres')
cleaned_titles['genres'].isnull().sum()

0

接下来评估`cleaned_credits`的缺失数据，其中只有`character`变量存在缺失值。  
角色名并不影响我们挖掘分析各流派中高IMDB评分作品演员，并且次变量缺失也可能因为演员类别是导演，没有对于角色，因此可以保留`character`变量存在空缺的观察值。

#### 处理重复数据

根据数据变量的含义以及内容来看，`cleaned_titles`里不应该存在每个变量值都相同的观察值，因此查看是否存在重复值。

In [84]:
cleaned_titles.duplicated().sum()

0

输出结果为零，说明不存在重复值  

接下来查看`cleaned_credits`数据表是否存在重复值。

In [85]:
cleaned_credits.duplicated().sum()

0

输出结果为零，说明不存在重复值

#### 处理不一致数据

针对`cleaned_titles`，不一致数据可能存在于`genres`和`character`变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。

In [86]:
cleaned_titles['genres'].value_counts().sum()

16970

从上面来看，`genres`列里并不存在不一致数据局，各个值都在代指不同的流派，但是里面还有空字符串表示的流派，并非是有效数据，应删除。  
删除后检查`genres`中是否还存在空字符串的行。

In [90]:
cleaned_titles = cleaned_titles.query('genres != ""')
cleaned_titles.query('genres == ""')

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

接下来，针对`production_countries`列也是一样的流程，利用`value_counts`方法，得到`production_countries`的列表里面各个值的出现次数。

In [91]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

由于`value_counts`执行结果中有太多值，Pandas只会默认显示开头和结尾的一些值。要完整展示结果，可以把`display.max_rows`设置为`None`，即取消展示行数上限。

但因为我们只是在当前调用`value_counts`时才需要看完整结果，所以可以结合`option_context`，只更改临时上限。

In [92]:
with pd.option_context('display.max_rows',None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

从输出结果来看，出品国家都用了两位的国家代码来表示，除了存在一个`lebanon`的值。  
`lebanon`的国家代码为`LB`，二者都在表示一个国家，需要进行统一。  
进行统一后并检查替换后的结果。  

In [96]:
# 先进行替换
cleaned_titles.replace('lebanon','LB',inplace=True)
# 在检查结果
cleaned_titles[cleaned_titles['production_countries'] == 'lebanon']

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

另外，里面还存在空字符串表示的国家代码，并非有效数据。但由于出品国家并非分析所需的关键信息，所以可以保留出品国家为空的观察值。

针对`original_credits`，不一致数据可能存在于`role`中，我们将查看是否存在多个不同值指代同一演职员类型的情况。

In [97]:
original_credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

从上面结果来看，`role`只有两种可能，不存在不一致数据，因此可以把这列的类型转换为`Category`，好处是比字符串更节约内存空间，也能说明值的类型。

In [98]:
cleaned_credits['role']  =cleaned_credits['role'].astype('category')
cleaned_credits['role']

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### 处理无效或错误数据

可以通过DateFrame的`describe`方法，对数值统计信息进行快速了解

In [99]:
original_titles.describe()

release_year      runtime      seasons   imdb_score    imdb_votes  \
count   5850.000000  5850.000000  2106.000000  5368.000000  5.352000e+03   
mean    2016.417094    76.888889     2.162868     6.510861  2.343938e+04   
std        6.937726    39.002509     2.689041     1.163826  9.582047e+04   
min     1945.000000     0.000000     1.000000     1.500000  5.000000e+00   
25%     2016.000000    44.000000     1.000000     5.800000  5.167500e+02   
50%     2018.000000    83.000000     1.000000     6.600000  2.233500e+03   
75%     2020.000000   104.000000     2.000000     7.300000  9.494000e+03   
max     2022.000000   240.000000    42.000000     9.600000  2.294231e+06   

       tmdb_popularity   tmdb_score  
count      5759.000000  5539.000000  
mean         22.637925     6.829175  
std          81.680263     1.170391  
min           0.009442     0.500000  
25%           2.728500     6.100000  
50%           6.821000     6.900000  
75%          16.590000     7.537500  
max        2274.044000    10.000000

从上述结果来看，`original_titles`不存在脱离现实意义的数值。  `original_titles`由于不包含表示数值含义的变量，所以无需用`describe`检查。

## 整理数据

In [100]:
cleaned_titles

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                            description release_year  \
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
...                                                 ...          ...   
5846  A heart warming film that explores the concept...   2021-01-01   
5847  A controversial TV host and comedian who has b...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [101]:
cleaned_credits

person_id         id                name                character  \
0          3748    tm84618      Robert De Niro            Travis Bickle   
1         14658    tm84618        Jodie Foster            Iris Steensma   
2          7064    tm84618       Albert Brooks                      Tom   
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4         48933    tm84618     Cybill Shepherd                    Betsy   
...         ...        ...                 ...                      ...   
77796    736339  tm1059008    Adelaida Buscato                María Paz   
77797    399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798    373198  tm1059008         Inés Prieto                    Fanny   
77799    378132  tm1059008        Isabel Gaona                   Cacica   
77800   1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

对数据的整理，与分析的方向紧密相关，此次数据分析的目的是，整理不同流派的影视作品，比如如喜剧片、动作片、科幻片中，演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。  
那为了能同时获得流派与演员数据，我们需要把`cleaned_titles`和`cleaned_credits`，通过`id`作为键进行连接，因为两个数据表中`id`都是影视作品ID,连接方式选择`inner`保证连接后的数据有意义。

In [107]:
credits_with_titles = pd.merge(cleaned_titles,cleaned_credits,on='id',how='inner')

连接后，就能知道各个演员参与过的影视的具体信息

In [108]:
credits_with_titles.head()

id        title   type  \
0  tm84618  Taxi Driver  MOVIE   
1  tm84618  Taxi Driver  MOVIE   
2  tm84618  Taxi Driver  MOVIE   
3  tm84618  Taxi Driver  MOVIE   
4  tm84618  Taxi Driver  MOVIE   

                                         description release_year  \
0  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  drama                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  drama                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score person_id  \
0         8.2    808582.0           40.965       8.179      3748   
1         8.2    808582.0           40.965       8.179     14658   
2         8.2    808582.0           40.965       8.179      7064   
3         8.2    808582.0           40.965       8.179      3739   
4         8.2    808582.0           40.965       8.179     48933   

              name                character   role  
0   Robert De Niro            Travis Bickle  ACTOR  
1     Jodie Foster            Iris Steensma  ACTOR  
2    Albert Brooks                      Tom  ACTOR  
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  
4  Cybill Shepherd                    Betsy  ACTOR

由于我们只对挖掘演员的参演作品口碑感兴趣，导演不在我们的分析范围内，因此根据`role`，筛选出类型为`ACTOR`的观察值，供后续分析。

In [109]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

为了挖掘出各个流派中高IMDB评分的演员，我们要先根据流派和演员进行分组。  
对演员进行分组的时候，选择的是用`person_id`而不是`name`变量原因是名字容易出现错拼或者重名的情况，演职员ID会比演员姓名更加准确地反映是哪位演员。

In [110]:
groupby_genres_and_person_id = actor_with_titles.groupby(['genres','person_id'])

分组后，我们只需要对`imdb_score`的进行聚合计算，因此只提取`imdb_score`变量，然后调用`mean`，计算各个影视流派作品中，每位演员参演作品的平均IMDB评分。

In [112]:
imdb_groupby_genres_and_person_id = groupby_genres_and_person_id['imdb_score'].mean()
imdb_groupby_genres_and_person_id 

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

我们可以调用`reset_index`，对层次化索引进行重置，得到更加规整的DataFrame

In [113]:
imdb_groupby_genres_and_person_id_df = imdb_groupby_genres_and_person_id.reset_index()
imdb_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

现在针对流派和演员分组的IMDB评分数据已经整理好，可以进入后续的分析步骤了。

但我们当前可以继续做一些数据整理，比如对上面的结果再次进行分组，找出各个流派里演员作品最高的平均评分是多少、最高评分对应的演员名字是什么。

要得到这一结果，我们需要再次用`genres`进行分组，然后提取出`imdb_score`变量，计算其最大值。

In [114]:
genres_max_scores = imdb_groupby_genres_and_person_id_df.groupby('genres')['imdb_score'].max()
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

在我们知道最高分后，可以把以上的结果和之前的结果再次进行连接，得到得分最高的各个演员ID是什么，也就是这个最高平均分是哪位演员得到的。

In [116]:
genres_max_scores_and_person_id_df = pd.merge(genres_max_scores,imdb_groupby_genres_and_person_id_df,on=['genres','imdb_score'])
genres_max_scores_and_person_id_df

genres  imdb_score person_id
0     action         9.3     12790
1     action         9.3      1303
2     action         9.3     21033
3     action         9.3    336830
4     action         9.3     86591
..       ...         ...       ...
131      war         8.8    826547
132  western         8.9     22311
133  western         8.9     28166
134  western         8.9     28180
135  western         8.9     93017

[136 rows x 3 columns]

从上面的结果可以看出，最高得分的对应的演员不一定只有一位，可能有多位演员的平均分相同。  
为了得到演员ID所对应的名字，我们可以和`cleaned_credits`这个DataFrame进行连接，这个DataFrame还有其他的列，我们只需其中`person_id`和`name`的对应，所以可以先提取出那两列，并把重复的行进行删除

In [118]:
actor_id_with_names = cleaned_credits[['person_id','name']].drop_duplicates()
actor_id_with_names

person_id                name
0          3748      Robert De Niro
1         14658        Jodie Foster
2          7064       Albert Brooks
3          3739       Harvey Keitel
4         48933     Cybill Shepherd
...         ...                 ...
77796    736339    Adelaida Buscato
77797    399499  Luz Stella Luengas
77798    373198         Inés Prieto
77799    378132        Isabel Gaona
77800   1950416      Julian Gaviria

[54589 rows x 2 columns]

下一步就可以把`actor_id_with_names`与前面得到的`genres_max_score_with_person_id`进行连接，增加`name`变量，从而展示平均评分最高的演员名字。

In [127]:
genres_max_scores_with_actor_name = pd.merge(genres_max_scores_and_person_id_df,actor_id_with_names,on='person_id')
genres_max_scores_with_actor_name

genres  imdb_score person_id                name
0       action         9.3     12790         Olivia Hack
1        scifi         9.3     12790         Olivia Hack
2       action         9.3      1303       Jessie Flower
3    animation         9.3      1303       Jessie Flower
4       family         9.3      1303       Jessie Flower
..         ...         ...       ...                 ...
131        war         8.8    826547         Yuto Uemura
132    western         8.9     22311     Koichi Yamadera
133    western         8.9     28166  Megumi Hayashibara
134    western         8.9     28180      Unsho Ishizuka
135    western         8.9     93017            Aoi Tada

[136 rows x 4 columns]

为了把相同的流派都排序在一起，我们还可以引用`sort_values`方法，把结果里面的行根据`genres`进行排序，然后用`reset_index`把索引重新排序。  
索引重新排序后，DataFrame会多出`index`一列，我们可以再把`index`列进行删除

In [128]:
genres_max_scores_with_actor_name = genres_max_scores_with_actor_name.sort_values("genres").reset_index().drop('index',axis=1)
genres_max_scores_with_actor_name

genres  imdb_score person_id                name
0     action         9.3     12790         Olivia Hack
1     action         9.3    336830     André Sogliuzzo
2     action         9.3     21033          Zach Tyler
3     action         9.3     86591       Cricket Leigh
4     action         9.3      1303       Jessie Flower
..       ...         ...       ...                 ...
131      war         8.8    826547         Yuto Uemura
132  western         8.9     28166  Megumi Hayashibara
133  western         8.9     28180      Unsho Ishizuka
134  western         8.9     22311     Koichi Yamadera
135  western         8.9     93017            Aoi Tada

[136 rows x 4 columns]